### DISCUSSION TITLE
Exception Handling in Instrumentation for GraphQL Custom RateLimiter per User

### DISCUSSION THREAD

**Discussion by sbilello** [graphql-java/graphql-java#2485] [General]
Hi everyone!

I was looking at this StackOverflow answer to handle Role Authorization via `Instrumentation` https://stackoverflow.com/questions/51764092/how-to-get-query-mutation-operation-name/51845557#comment98112574_51845557

I would love to create a `graphql-rate-limiter` for each field and for a particular user

Something like:

```
@Component
@Slf4j
public class RateLimitInstrumentation extends SimpleInstrumentation {

    @Override
    public DataFetcher<?> instrumentDataFetcher(DataFetcher<?> dataFetcher, InstrumentationFieldFetchParameters parameters) {
        DgsContext dgsContext = parameters.getEnvironment().getContext();
        ServletWebRequest request = (ServletWebRequest) dgsContext.getRequestData().getWebRequest();
        Object object =  request.getAttribute(Constants.USER_CONFIG, ServletWebRequest.SCOPE_REQUEST);
        GraphQLFieldDefinition fieldDefinition = parameters.getEnvironment().getFieldDefinition();
        log.info("GraphQLFieldDefinition {}", fieldDefinition.getName());
        if (object != null && object instanceof UserConfig) {
            UserConfig userConfig = (UserConfig) object;
           //  TODO Implement rate limit logic 
            throw new RateLimitException("over the threshold for userConfig " + userConfig.getRateLimitConfig());
        }
        else return super.instrumentDataFetcher(dataFetcher, parameters);
    }
}
```

This approach can work but I don't like that It bypasses https://github.com/graphql-java/graphql-java/blob/master/src/main/java/graphql/execution/ExecutionStrategy.java#L356

Does it make sense to wrap under `try and catch` the
https://github.com/graphql-java/graphql-java/blob/master/src/main/java/graphql/execution/ExecutionStrategy.java#L275
```        dataFetcher = instrumentation.instrumentDataFetcher(dataFetcher, instrumentationFieldFetchParams);```
so we can delegate the exception handling?
Maybe we should also cover with `try and catch`:
```InstrumentationContext<Object> fetchCtx = instrumentation.beginFieldFetch(instrumentationFieldFetchParams);```

Note: I use an interceptor to populate the request with the `UserConfig` attribute. If you can advise some other approach please let me know. I wanted to avoid using ThreadLocal variables and make my fork of the `graphql-java` library. :)

Thanks

**Comment by kaqqao** [comment]
There's a key difference between your code and the example on StackOverflow. You throw an exception in the instrumentation, whereas I produce a `DataFetcher` that throws an exception (which then gets handled like any other exception thrown inside a `DataFetcher`). I guess this is what you mean by the usual exception handling being skipped.

**Reply by sbilello** [reply]
```
//Checks if the current user has the needed roles for each field
public class AuthInstrumentation extends SimpleInstrumentation {
    @Override
    public DataFetcher<?> instrumentDataFetcher(DataFetcher<?> dataFetcher, InstrumentationFieldFetchParameters parameters) {
        GraphQLFieldDefinition fieldDefinition = parameters.getEnvironment().getFieldDefinition();
        //Each protected field is expected to have a directive called "auth" with an argument called "rolesRequired" that is a list of strings representing the roles
        Optional<GraphQLArgument> rolesRequired = DirectivesUtil.directiveWithArg(fieldDefinition.getDirectives(), "auth", "rolesRequired");
        if (rolesRequired.isPresent()) {
            List<String> roles = (List<String>) rolesRequired.get().getValue();
            User currentUser = parameters.getEnvironment().getContext(); //get the user from context
            if (!currentUser.getRoles().containsAll(roles)) {
                //Replace the normal resolution logic with the one that always returns null (or throws an exception) when the user doesn't have access
                return env -> null;
            }
        }
        return super.instrumentDataFetcher(dataFetcher, parameters);
    }
}
```
@kaqqao thanks for replying.
I saw the comment or `throws an exception` and this is what I followed. It is true that it returns `env->null` but I followed the code comment. I was wondering how should it work in that example in a graceful manner if it returns null. Of course, in the stackoverflow the caller would throw an NPE and it will be handled later on but it does not look to be handled gracefully. Am I missing something?

Another way on how I would solve the problem would be to avoid the instrumentation and maybe write some logic inside the datafetcher for the particular field and throw the exception from inside the datafetcher so we would leverage the `DataFetcherExceptionHandler`.

Or maybe I would create a custom `@GraphQLRateLimit` annotation that decorates the datafetcher

I am bouncing off some ideas.

**Reply by kaqqao** [reply]
Huh, the massage I posted above somehow got truncated. I could swear I typed another paragraph. Anyway, replace `throw new RateLimitException` with `return env -> throw new RateLimitException`. So instead of throwing an exception in the instrumentation, you replace the regular `DataFetcher` with one that always throws. That's what I meant with that comment.
Also, you can also push down the check into the generated DataFetcher if you prefer.

**Reply by sbilello** [reply]
Thanks @kaqqao 
I solved with the syntax ` return c -> {throw new RateLimitException("over the threshold");};`
It looks to be a little bit ugly but it works. I guess that if we move `dataFetcher = instrumentation.instrumentDataFetcher(dataFetcher, instrumentationFieldFetchParams); inside 
`try and catch` we can avoid such a workaround.
  
```
@Component
@Slf4j
public class RateLimitInstrumentation extends SimpleInstrumentation {

    @Override
    public DataFetcher<?> instrumentDataFetcher(DataFetcher<?> dataFetcher, InstrumentationFieldFetchParameters parameters) {
        DgsContext dgsContext = parameters.getEnvironment().getContext();
        ServletWebRequest request = (ServletWebRequest) dgsContext.getRequestData().getWebRequest();
        Object object =  request.getAttribute(Constants.USER_CONFIG, ServletWebRequest.SCOPE_REQUEST);
        GraphQLFieldDefinition fieldDefinition = parameters.getEnvironment().getFieldDefinition();
        log.info("GraphQLFieldDefinition {}", fieldDefinition.getName());
        if (object != null) {
            UserConfig userConfig = (UserConfig) object;
            // TODO IMPLEMENT RATE LIMITING LOGIC 
            return c -> {throw new RateLimitException("over the threshold");};
        }
        else return super.instrumentDataFetcher(dataFetcher, parameters);
    }
}
```
